### PR TITLE
[dagster-dbt] update usage of deprecated context methods

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
Some methods in the dbt integration use context methods that were deprecated in https://github.com/dagster-io/dagster/pull/19441 and https://github.com/dagster-io/dagster/pull/17339. The op-specific methods are meant to be replaced by calling `context.op_execution_context.op` rather than `context.op`. However, in some of the dbt methods, we don't know if the context is of type `AssetExecutionContext` or `OpExecutionContext`, so it is unclear if we need to call the underlying `op_execution_context` before op-specific methods. 

This PR adds a check in some methods that will pull the `op_execution_context` from the context if the context is an `AssetExecutionContext` so that we are guaranteed to be working with an `OpExecutionContext` in the rest of the method.

### CODE REVIEW

**Comment by jamiedemaria** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #19806** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19806?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19806?utm_source=stack-comment).

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 1]
```diff
@@ -967,9 +968,12 @@ def my_dbt_op(dbt: DbtCliResource):
                     dbt_macro_args = {"key": "value"}
                     dbt.cli(["run-operation", "my-macro", json.dumps(dbt_macro_args)]).wait()
         """
+        op_context = (
+            context.op_execution_context if isinstance(context, AssetExecutionContext) else context
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:972]: Can we update the type of `context` in the function to be a `Union`?

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 2]
```diff
@@ -967,9 +968,12 @@ def my_dbt_op(dbt: DbtCliResource):
...
+        )
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:973]: Would prefer that we just keep this named `context`. 

Can we move this to be after we instantiate `assets_def`, but before the `if context and assets_def is not None`?

```suggestion
        context: Optional[OpExecutionContext] = (
            context.op_execution_context if isinstance(context, AssetExecutionContext) else context
        )
```

Comment (jamiedemaria) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:973]: sure for the rename, but `_get_unique_target_name` needs the op context too and is called before we get the assets_def. I just moved fetching `assets_def` up, but lmk if you'd prefer something else

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 3]
```diff
@@ -969,6 +970,14 @@ def my_dbt_op(dbt: DbtCliResource):
         """
         dagster_dbt_translator = validate_opt_translator(dagster_dbt_translator)
 
+        assets_def: Optional[AssetsDefinition] = None
+        with suppress(DagsterInvalidPropertyError):
+            assets_def = context.assets_def if context else None
+
+        context: Optional[OpExecutionContext] = (
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:977]: I think if you remove the type annotation we'll pass pyright

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 4]
```diff
@@ -841,7 +842,7 @@ def cli(
         raise_on_error: bool = True,
         manifest: Optional[DbtManifestParam] = None,
         dagster_dbt_translator: Optional[DagsterDbtTranslator] = None,
-        context: Optional[OpExecutionContext] = None,
+        context: Optional[Union[OpExecutionContext, AssetExecutionContext]] = None,
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:845]: don't forget the docstring

**Review by rexledesma — APPROVED** [review]
Looks good once tests pass

**Pull request merged**